File: frontend/src/api/quality_loop.ts
import request from './request'

export interface LoopRule {
  id: number
  name: string
  trigger_metric: string
  threshold: number
  operator: string
  action_chain: Record<string, any> | null
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface LoopExecution {
  id: number
  rule_id: number
  rule_name: string | null
  trigger_value: number
  current_step: number
  total_steps: number
  status: string
  steps_log: { steps: { step: number; action: string; status: string; detail: string }[] } | null
  defect_id: number | null
  started_at: string
  completed_at: string | null
}

export function getLoopRules() {
  return request.get('/quality-loop/rules') as Promise<LoopRule[]>
}

export function createLoopRule(data: Partial<LoopRule>) {
  return request.post('/quality-loop/rules', data) as Promise<LoopRule>
}

export function updateLoopRule(id: number, data: Partial<LoopRule>) {
  return request.put(`/quality-loop/rules/${id}`, data) as Promise<LoopRule>
}

export function deleteLoopRule(id: number) {
  return request.delete(`/quality-loop/rules/${id}`) as Promise<{ ok: boolean }>
}

export function triggerRule(id: number) {
  return request.post(`/quality-loop/rules/${id}/trigger`) as Promise<LoopExecution>
}

export function getLoopExecutions(params?: { rule_id?: number; status?: string }) {
  return request.get('/quality-loop/executions', { params }) as Promise<LoopExecution[]>
}

export function evaluateRules() {
  return request.post('/quality-loop/evaluate') as Promise<{ triggered: number }>
}

File: backend/app/schemas/quality_loop.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict


class QualityLoopRuleBase(BaseModel):
    name: str
    trigger_metric: str
    threshold: float
    operator: str = "<"
    action_chain: dict | None = None
    enabled: bool = True


class QualityLoopRuleCreate(QualityLoopRuleBase):
    pass


class QualityLoopRuleUpdate(BaseModel):
    name: str | None = None
    trigger_metric: str | None = None
    threshold: float | None = None
    operator: str | None = None
    action_chain: dict | None = None
    enabled: bool | None = None


class QualityLoopRuleResponse(QualityLoopRuleBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    created_at: datetime
    updated_at: datetime


class QualityLoopExecutionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    rule_id: int
    rule_name: str | None = None
    trigger_value: float
    current_step: int
    total_steps: int
    status: str
    steps_log: dict | None = None
    defect_id: int | None = None
    started_at: datetime
    completed_at: datetime | None = None
